apache/nifi
# NIFI-5176 Java 11 Build Compatibilty

NiFi can be built and run on Java 11.

The goal of this PR is to establish general build compatibility for Java 11 while retaining complete build and runtime validity for Java 8.

TODOs/needed fixes will be added here.

Current status:

- [x] NiFi can be built on Java 11
- [x] NiFi starts successfully on Java 11
- [ ] Several standard processors have been tested, and work.  Should an exhaustive list be supplied?
- [x] Provenance events displaying properly in the Provenance UI
- [ ] Processors using Ignite dependencies will most likely not work on Java 11.  A release of the Ignite dependencies compatible with Java 11 has not yet been made
- [ ] JMS processors may not work.  ActiveMQ 5.15.8 does not run on Java 11.  Compatibility with Java 11 is expected in release version 5.16.0 of ActiveMQ.
- [x] Tests that use TLS often fail when specifying a protocal version of `TLS` or `TLSv1.3` in tests during the Java 11 build, need to investigate and resolve in a follow-on PR. For now, the tests explicitly use `TLSv1.2`.

Things to test with this PR:
- Building on Java 8 and running on Java 8
- Building on Java 8 and running on Java 11
- Building on Java 11 and running on Java 11

Note: When building, to work around `TestJdbcCommon` timezone issues (which also occur on master), add `-Duser.timezone=UTC` to the mvn command when building.

For the three build/run scenarios above:
- Verify running securely with certs generated by the TLS Toolkit
- Verify functionality with reverse proxies
- Verify processors, controller services, reporting tasks work
- Use TLSv1.2 explicitly

The goal of this PR isn't to fix _every_ issue running on Java 11, but it should not introduce any new issues for building on Java 8 and running the Java 8 build on Java 8 or 11.

Follow-on PRs can be created to fix issues encountered while running on Java 11 for specific components.
